commit 993a857ec1b32e3cd917020231e46bc502226960
Author: David S. Miller <davem@davemloft.net>
Date:   Wed Dec 8 18:42:23 2010 -0800

    econet: Fix crash in aun_incoming().
    
    Unconditional use of skb->dev won't work here,
    try to fetch the econet device via skb_dst()->dev
    instead.
    
    Suggested by Eric Dumazet.
    
    Reported-by: Nelson Elhage <nelhage@ksplice.com>
    Tested-by: Nelson Elhage <nelhage@ksplice.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>
    [dannf: adjusted to apply to Debian's 2.6.26]

diff --git a/net/econet/af_econet.c b/net/econet/af_econet.c
index 70a161f..745e4c6 100644
--- a/net/econet/af_econet.c
+++ b/net/econet/af_econet.c
@@ -847,9 +847,13 @@ static void aun_incoming(struct sk_buff *skb, struct aunhdr *ah, size_t len)
 {
 	struct iphdr *ip = ip_hdr(skb);
 	unsigned char stn = ntohl(ip->saddr) & 0xff;
+	struct dst_entry *dst = skb_dst(skb);
+	struct ec_device *edev = NULL;
 	struct sock *sk;
 	struct sk_buff *newskb;
-	struct ec_device *edev = skb->dev->ec_ptr;
+
+	if (dst)
+		edev = dst->dev->ec_ptr;
 
 	if (! edev)
 		goto bad;
